pkg/server/fsm.go

```diff
@@ -1074,7 +1075,16 @@ func (h *fsmHandler) recvMessageWithError() (*fsmMsg, error) {
 				h.fsm.lock.RLock()
 				rfMap := h.fsm.rfMap
 				h.fsm.lock.RUnlock()
-				ok, err := bgp.ValidateUpdateMsg(body, rfMap, isEBGP, isConfed)
+
+				// Allow updates from loopback addresses if the GoBGP instance
+				// itself is assigned to 127.0.0.0/8, since this can happen when
```

Comment:
What `assigned` mean?

Reply:
"assigned" means that the GoBGP server is expected set the next-hop attribute to one of these addresses, which generally should not happen.

The use case I have is I'm starting multiple GoBGP servers on the same host on different loopback addresses for lightweight testing. In order for this to work, I need the UPDATE messages to not be dropped when the next-hop is set to 127.0.0.0/8. I set this up by configuring Neighbor.Transport.Config.LocalAddress to the specific IP address within 127.0.0.0/8 for each test GoBGP instance.

In order to distinguish this test scenario from production scenarios, I thought the right way would be to check whether the RouterID falls within 127/8.

I wasn't able to figure out how GoBGP exactly sets the nexthop attribute, which should never be 127/8 in production use cases. Do you think there is a better way to check for this to make GoBGP more testable?

Reply:
You can set RouterID to anything so I don't think that it's the right way to check if nexthops points to a loopback address or not.
Why not using `fsm.peerInfo.Address` or `fsm.peerInfo.LocalAddress`?

The following code updates the next hop attribute:
https://github.com/osrg/gobgp/blob/146b2b8f598d5f5c1c674fa451a2ccf0781220bf/internal/pkg/table/path.go#L197